### PR TITLE
[libcpuid] Add new port

### PR DESCRIPTION
**Pull request by kreuzerkrieg**
Add libcpuid port

- Which triplets are supported/not supported? 
Theoretically all except UWP.
- Have you updated the CI baseline?
Yes
- Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
Yes

### CODE REVIEW

**Comment by LilyWangL** [comment]
@kreuzerkrieg, currently, this port build failed on Windows with the following error:
```
[11/12] cmd.exe /C "cd . && D:\downloads\tools\cmake-3.14.0-windows\cmake-3.14.0-win32-x86\bin\cmake.exe -E vs_link_exe --intdir=cpuid_tool\CMakeFiles\cpuid_tool.dir --rc="C:\PROGRA~2\Windows Kits\10\bin\10.0.18362.0\x64\rc.exe" --mt="C:\PROGRA~2\Windows Kits\10\bin\10.0.18362.0\x64\mt.exe" --manifests  -- C:\VS2017\Stable\VC\Tools\MSVC\14.22.27905\bin\Hostx64\x64\link.exe  cpuid_tool\CMakeFiles\cpuid_tool.dir\cpuid_tool.c.obj  /out:cpuid_tool\cpuid_tool.exe /implib:cpuid_tool\cpuid_tool.lib /pdb:cpuid_tool\cpuid_tool.pdb /version:0.0  /machine:x64 /debug /INCREMENTAL /subsystem:console  libcpuid\cpuid.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ."
FAILED: cpuid_tool/cpuid_tool.exe 
cmd.exe /C "cd . && D:\downloads\tools\cmake-3.14.0-windows\cmake-3.14.0-win32-x86\bin\cmake.exe -E vs_link_exe --intdir=cpuid_tool\CMakeFiles\cpuid_tool.dir --rc="C:\PROGRA~2\Windows Kits\10\bin\10.0.18362.0\x64\rc.exe" --mt="C:\PROGRA~2\Windows Kits\10\bin\10.0.18362.0\x64\mt.exe" --manifests  -- C:\VS2017\Stable\VC\Tools\MSVC\14.22.27905\bin\Hostx64\x64\link.exe  cpuid_tool\CMakeFiles\cpuid_tool.dir\cpuid_tool.c.obj  /out:cpuid_tool\cpuid_tool.exe /implib:cpuid_tool\cpuid_tool.lib /pdb:cpuid_tool\cpuid_tool.pdb /version:0.0  /machine:x64 /debug /INCREMENTAL /subsystem:console  libcpuid\cpuid.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ."
LINK Pass 1: command "C:\VS2017\Stable\VC\Tools\MSVC\14.22.27905\bin\Hostx64\x64\link.exe cpuid_tool\CMakeFiles\cpuid_tool.dir\cpuid_tool.c.obj /out:cpuid_tool\cpuid_tool.exe /implib:cpuid_tool\cpuid_tool.lib /pdb:cpuid_tool\cpuid_tool.pdb /version:0.0 /machine:x64 /debug /INCREMENTAL /subsystem:console libcpuid\cpuid.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTFILE:cpuid_tool\CMakeFiles\cpuid_tool.dir/intermediate.manifest cpuid_tool\CMakeFiles\cpuid_tool.dir/manifest.res" failed (exit code 1120) with the following output:
Microsoft (R) Incremental Linker Version 14.22.27905.0
Copyright (C) Microsoft Corporation.  All rights reserved.

cpuid.lib(cpuid_main.c.obj) : error LNK2019: unresolved external symbol exec_cpuid referenced in function cpu_exec_cpuid
cpuid.lib(rdtsc.c.obj) : error LNK2019: unresolved external symbol cpu_rdtsc referenced in function cpu_tsc_mark
cpuid.lib(rdtsc.c.obj) : error LNK2019: unresolved external symbol busy_sse_loop referenced in function cpu_clock_by_ic
cpuid_tool\cpuid_tool.exe : fatal error LNK1120: 3 unresolved externals
```

**Comment by kreuzerkrieg** [comment]
> @kreuzerkrieg, currently, this port build failed on Windows with the following error:
> 
> ```
> [11/12] cmd.exe /C "cd . && D:\downloads\tools\cmake-3.14.0-windows\cmake-3.14.0-win32-x86\bin\cmake.exe -E vs_link_exe --intdir=cpuid_tool\CMakeFiles\cpuid_tool.dir --rc="C:\PROGRA~2\Windows Kits\10\bin\10.0.18362.0\x64\rc.exe" --mt="C:\PROGRA~2\Windows Kits\10\bin\10.0.18362.0\x64\mt.exe" --manifests  -- C:\VS2017\Stable\VC\Tools\MSVC\14.22.27905\bin\Hostx64\x64\link.exe  cpuid_tool\CMakeFiles\cpuid_tool.dir\cpuid_tool.c.obj  /out:cpuid_tool\cpuid_tool.exe /implib:cpuid_tool\cpuid_tool.lib /pdb:cpuid_tool\cpuid_tool.pdb /version:0.0  /machine:x64 /debug /INCREMENTAL /subsystem:console  libcpuid\cpuid.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ."
> FAILED: cpuid_tool/cpuid_tool.exe 
> cmd.exe /C "cd . && D:\downloads\tools\cmake-3.14.0-windows\cmake-3.14.0-win32-x86\bin\cmake.exe -E vs_link_exe --intdir=cpuid_tool\CMakeFiles\cpuid_tool.dir --rc="C:\PROGRA~2\Windows Kits\10\bin\10.0.18362.0\x64\rc.exe" --mt="C:\PROGRA~2\Windows Kits\10\bin\10.0.18362.0\x64\mt.exe" --manifests  -- C:\VS2017\Stable\VC\Tools\MSVC\14.22.27905\bin\Hostx64\x64\link.exe  cpuid_tool\CMakeFiles\cpuid_tool.dir\cpuid_tool.c.obj  /out:cpuid_tool\cpuid_tool.exe /implib:cpuid_tool\cpuid_tool.lib /pdb:cpuid_tool\cpuid_tool.pdb /version:0.0  /machine:x64 /debug /INCREMENTAL /subsystem:console  libcpuid\cpuid.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ."
> LINK Pass 1: command "C:\VS2017\Stable\VC\Tools\MSVC\14.22.27905\bin\Hostx64\x64\link.exe cpuid_tool\CMakeFiles\cpuid_tool.dir\cpuid_tool.c.obj /out:cpuid_tool\cpuid_tool.exe /implib:cpuid_tool\cpuid_tool.lib /pdb:cpuid_tool\cpuid_tool.pdb /version:0.0 /machine:x64 /debug /INCREMENTAL /subsystem:console libcpuid\cpuid.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTFILE:cpuid_tool\CMakeFiles\cpuid_tool.dir/intermediate.manifest cpuid_tool\CMakeFiles\cpuid_tool.dir/manifest.res" failed (exit code 1120) with the following output:
> Microsoft (R) Incremental Linker Version 14.22.27905.0
> Copyright (C) Microsoft Corporation.  All rights reserved.
> 
> cpuid.lib(cpuid_main.c.obj) : error LNK2019: unresolved external symbol exec_cpuid referenced in function cpu_exec_cpuid
> cpuid.lib(rdtsc.c.obj) : error LNK2019: unresolved external symbol cpu_rdtsc referenced in function cpu_tsc_mark
> cpuid.lib(rdtsc.c.obj) : error LNK2019: unresolved external symbol busy_sse_loop referenced in function cpu_clock_by_ic
> cpuid_tool\cpuid_tool.exe : fatal error LNK1120: 3 unresolved externals
> ```

Looks like x64 failed, will check it. Whats about x86 and arm?

**Comment by kreuzerkrieg** [comment]
> @kreuzerkrieg, currently, this port build failed on Windows with the following error:
> 
> ```
> [11/12] cmd.exe /C "cd . && D:\downloads\tools\cmake-3.14.0-windows\cmake-3.14.0-win32-x86\bin\cmake.exe -E vs_link_exe --intdir=cpuid_tool\CMakeFiles\cpuid_tool.dir --rc="C:\PROGRA~2\Windows Kits\10\bin\10.0.18362.0\x64\rc.exe" --mt="C:\PROGRA~2\Windows Kits\10\bin\10.0.18362.0\x64\mt.exe" --manifests  -- C:\VS2017\Stable\VC\Tools\MSVC\14.22.27905\bin\Hostx64\x64\link.exe  cpuid_tool\CMakeFiles\cpuid_tool.dir\cpuid_tool.c.obj  /out:cpuid_tool\cpuid_tool.exe /implib:cpuid_tool\cpuid_tool.lib /pdb:cpuid_tool\cpuid_tool.pdb /version:0.0  /machine:x64 /debug /INCREMENTAL /subsystem:console  libcpuid\cpuid.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ."
> FAILED: cpuid_tool/cpuid_tool.exe 
> cmd.exe /C "cd . && D:\downloads\tools\cmake-3.14.0-windows\cmake-3.14.0-win32-x86\bin\cmake.exe -E vs_link_exe --intdir=cpuid_tool\CMakeFiles\cpuid_tool.dir --rc="C:\PROGRA~2\Windows Kits\10\bin\10.0.18362.0\x64\rc.exe" --mt="C:\PROGRA~2\Windows Kits\10\bin\10.0.18362.0\x64\mt.exe" --manifests  -- C:\VS2017\Stable\VC\Tools\MSVC\14.22.27905\bin\Hostx64\x64\link.exe  cpuid_tool\CMakeFiles\cpuid_tool.dir\cpuid_tool.c.obj  /out:cpuid_tool\cpuid_tool.exe /implib:cpuid_tool\cpuid_tool.lib /pdb:cpuid_tool\cpuid_tool.pdb /version:0.0  /machine:x64 /debug /INCREMENTAL /subsystem:console  libcpuid\cpuid.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ."
> LINK Pass 1: command "C:\VS2017\Stable\VC\Tools\MSVC\14.22.27905\bin\Hostx64\x64\link.exe cpuid_tool\CMakeFiles\cpuid_tool.dir\cpuid_tool.c.obj /out:cpuid_tool\cpuid_tool.exe /implib:cpuid_tool\cpuid_tool.lib /pdb:cpuid_tool\cpuid_tool.pdb /version:0.0 /machine:x64 /debug /INCREMENTAL /subsystem:console libcpuid\cpuid.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTFILE:cpuid_tool\CMakeFiles\cpuid_tool.dir/intermediate.manifest cpuid_tool\CMakeFiles\cpuid_tool.dir/manifest.res" failed (exit code 1120) with the following output:
> Microsoft (R) Incremental Linker Version 14.22.27905.0
> Copyright (C) Microsoft Corporation.  All rights reserved.
> 
> cpuid.lib(cpuid_main.c.obj) : error LNK2019: unresolved external symbol exec_cpuid referenced in function cpu_exec_cpuid
> cpuid.lib(rdtsc.c.obj) : error LNK2019: unresolved external symbol cpu_rdtsc referenced in function cpu_tsc_mark
> cpuid.lib(rdtsc.c.obj) : error LNK2019: unresolved external symbol busy_sse_loop referenced in function cpu_clock_by_ic
> cpuid_tool\cpuid_tool.exe : fatal error LNK1120: 3 unresolved externals
> ```

nm, I think I've fixed it

**Comment by kreuzerkrieg** [comment]
ok, both Win x86 and x64 building on my machine, so I guess, now the ARM is failing?

**Comment by LilyWangL** [comment]
Now this port build failed on triplet x64-windows/x64-windows-static/arm64-windows. You can click the CI test details to get more information about test result.

**Comment by kreuzerkrieg** [comment]
> Now this port build failed on triplet x64-windows/x64-windows-static/arm64-windows. You can click the CI test details to get more information about test result.

wow! a year ago (AFAIR) there was no public access to build results. Now it makes life much easier. 
Fixed.

**Review by LilyWangL — APPROVED** [review]

**Comment by LilyWangL** [comment]
LGTM, thanks for this PR!

File: ports/libcpuid/lib.cmake [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+target_include_directories(cpuid SYSTEM PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>)
+
+target_compile_definitions(cpuid PRIVATE VERSION="${PROJECT_VERSION}")
+set(CPUID_HEADERS ${CMAKE_CURRENT_SOURCE_DIR}/libcpuid.h ${CMAKE_CURRENT_SOURCE_DIR}/libcpuid_constants.h ${CMAKE_CURRENT_SOURCE_DIR}/libcpuid_types.h)
+set_target_properties(cpuid PROPERTIES
+                      PUBLIC_HEADER "${CPUID_HEADERS}"
+                      )
+
+#Documentation
+find_package(Doxygen)
```
Comment (ras0219-msft) [inline ports/libcpuid/lib.cmake]: We conventionally don't build documentation for packages; we should ideally not include this in a buildsystem replacement.

At a minimum, we should disable it via passing `-DENABLE_DOCS=OFF` in configure.

**Review by ras0219-msft — CHANGES_REQUESTED** [review]

File: ports/libcpuid/lib.cmake (hunk 1 above)
Comment (kreuzerkrieg) [reply ports/libcpuid/lib.cmake]: Done

**Comment by kreuzerkrieg** [comment]
Hm... no idea how and why failure in `qt5` (suddenly?) fails the test. Did I get it right?

**Review by NancyLi1013 — APPROVED** [review]

File: scripts/ci.baseline.txt [hunk 2]
```diff
@@ -1554,8 +1557,8 @@ sdl2-mixer:arm-uwp=fail
 sdl2-mixer:x64-uwp=fail
 sdl2-net:arm-uwp=fail
 sdl2-net:x64-uwp=fail
-seal:arm-uwp=fail
-seal:x64-uwp=fail
+seal:arm-uwp=fail
```
Comment (strega-nil) [inline scripts/ci.baseline.txt]: what happened here?

File: ports/libcpuid/CONTROL [hunk 3]
```diff
@@ -0,0 +1,4 @@
+Source: libcpuid
+Version: 0.4.1
+Homepage: https://github.com/anrieff/libcpuid
+Description: Provides CPU identification for the x86 (and x86_64)
```
Comment (strega-nil) [inline ports/libcpuid/CONTROL]: Could you add a `Supports: x86 | x64` line?

Comment (kreuzerkrieg) [reply ports/libcpuid/CONTROL]: Done

File: scripts/ci.baseline.txt (hunk 2 above)
Comment (kreuzerkrieg) [reply scripts/ci.baseline.txt]: EOL I guess, will try to revert

Comment (kreuzerkrieg) [reply scripts/ci.baseline.txt]: Done

**Comment by strega-nil** [comment]
Alright, looks good to me! Thanks @kreuzerkrieg :D

**Pull request merged**